### PR TITLE
Fix test 4270

### PR DESCRIPTION
**Pull request by qrilka**
Please include the following checklist in your PR:

* [ ] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [ ] The documentation has been updated, if necessary.

This is a fix for a test `4270-files-order`, the main point of it is in af36e04 other commits improve Stack integration tests a bit.
Integration tests succeed locally.

### CODE REVIEW

File: src/Stack/Package.hs [hunk 1]
```diff
@@ -803,12 +803,22 @@ resolveComponentFiles component build names = do
...
+    csources <-
+        forMaybeM (cSources build) $ \fp -> do
+            result <- resolveDirFile dir fp
+            case result of
+              Nothing -> do
+                warnMissingFile "File" cwd fp file
+                return Nothing
+              Just p -> return $ Just (DotCabalCFilePath p)
+    jsources <-
+        forMaybeM (targetJsSources build) $ \fp ->
```
Comment (snoyberg) [inline src/Stack/Package.hs]: Does it make sense to use `findCandidate` for the Javascript source? If so, why? Why not use the same logic as the C sources?

Comment (qrilka) [reply src/Stack/Package.hs]: oh, that's just a stupid miss from me, will fix that

File: ChangeLog.md [hunk 2]
```diff
@@ -97,6 +97,8 @@ Bug fixes:
   [#4314](https://github.com/commercialhaskell/stack/pull/4314)
 * Add `--cabal-files` flag to `stack ide targets` command.
 * Don't download ghc when using `stack clean`.
+* Support loading in GHCi definitions from symlinked C files. See
```
Comment (snoyberg) [inline ChangeLog.md]:
```suggestion
* Support loading in GHCi definitions from symlinked C files. Without this patch, Stack will try to find object files in the directory pointed to by symlinks, while GCC will produce the object files in the original directory. See
```

File: src/Stack/Package.hs [hunk 3]
```diff
@@ -1227,14 +1236,16 @@ findCandidate dirs name = do
...
-        :: (MonadIO m, MonadThrow m)
-        => Path Abs Dir -> FilePath.FilePath -> m [Path Abs File]
-    resolveCandidate x y = do
-        -- The standard canonicalizePath does not work for this case
-        p <- parseCollapsedAbsFile (toFilePath x FilePath.</> y)
-        exists <- doesFileExist p
-        return $ if exists then [p] else []
+    resolveCandidate dir = fmap maybeToList . resolveDirFile dir
+
+resolveDirFile
```
Comment (snoyberg) [inline src/Stack/Package.hs]: Please add a comment explaining that symlinks will not be followed.

**Comment by qrilka** [comment]
Anything else here?

**Comment by snoyberg** [comment]
Good to merge after CI runs

**Pull request merged**